tests/performance/density.go

```diff
@@ -0,0 +1,186 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package performance
+
+import (
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+
+	kvv1 "kubevirt.io/client-go/api/v1"
+	cd "kubevirt.io/kubevirt/tests/containerdisk"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/util"
+)
+
+var _ = SIGDescribe("Control Plane Performance Density Testing", func() {
+	var (
+		err        error
+		virtClient kubecli.KubevirtClient
+	)
+
+	BeforeEach(func() {
+		virtClient, err = kubecli.GetKubevirtClient()
+		util.PanicOnError(err)
+		tests.BeforeTestCleanup()
+	})
+
+	// Run all density tests
+	Describe("Density test", func() {
+		vmCount := 100
+		vmBatchStartupLimit := 300 * time.Second
+
+		Context(fmt.Sprintf("[small] create a batch of %d VMIs", vmCount), func() {
+
+			It("should sucessfully create all VMIS", func() {
+				quit := make(chan bool)
+				runningVMI := make(map[string]int)
+
+				By("Starting vmi watcher")
+				go watchVMI(virtClient, vmBatchStartupLimit, quit, &runningVMI)
+
+				By("Creating a batch of VMIs")
+				createBatchVMIWithRateControl(virtClient, vmCount)
+
+				By("Waiting a batch of VMIs")
+				waitRunningVMI(&runningVMI, vmCount, vmBatchStartupLimit)
+
+				By("Deleting all VMIs")
+				deleteVMIs(virtClient, runningVMI)
+
+				By("Stop vmi watcher")
+				close(quit)
+			})
+		})
+	})
+})
+
+// createBatchVMIWithRateControl creates a batch of vms concurrently, uses one goroutine for each creation.
+// between creations there is an interval for throughput control
+func createBatchVMIWithRateControl(virtClient kubecli.KubevirtClient, vmCount int) {
+	defer GinkgoRecover()
+	for i := 1; i <= vmCount; i++ {
+		vmi := createVMISpecWithResources(virtClient)
+		By(fmt.Sprintf("Creating VMI %s", vmi.ObjectMeta.Name))
+		_, err := virtClient.VirtualMachineInstance(util.NamespaceTestDefault).Create(vmi)
+		Expect(err).ToNot(HaveOccurred())
+
+		// sleep to manually throttle API requests.
+		time.Sleep(300 * time.Millisecond)
```

Comment:
Do you try to protect  the API with that? If only that is the use-case, the value seems to be very high.

Reply:
Right, the legend is not accurate. This sleep is to control the creation rate.
This is the same as in the Kubernetes density test [here](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/density_test.go#L349)